pkg/cloudprovider/types.go

```diff
@@ -257,6 +257,31 @@ func (ofs Offerings) Cheapest() Offering {
 	})
 }
 
+// MostExpensive returns the most expensive offering from the return offerings
+func (ofs Offerings) MostExpensive() Offering {
+	return lo.MaxBy(ofs, func(a, b Offering) bool {
+		return a.Price > b.Price
+	})
+}
+
+// CapacityType returns the capacity type requirement for an offering, and panics if a capacity type is not set
+func (of Offering) CapacityType() string {
+	req, ok := of.Requirements[v1beta1.CapacityTypeLabelKey]
+	if !ok {
+		panic("failed to get capacity type for offering, capacity type must be definied on all offerings")
+	}
```

Comment:
Does this need to be a panic? This should also be an invariant enforced across our code, right? Can this ever happen? 

Can we do a fatal logging instead of a panic? Maybe even just an error log?

Reply:
This should be an invariant, maybe a fatal log is the better approach. Could also make it an error, but it was a programming mistake if this isn't present, not a runtime error.

Reply:
How much do we need these helper functions? Is it a lot to grab the scheduling requirement back from the offering directly?

Reply:
Interesting enough: It seems like these functions are really only used in testing and I think there is one usage in the kwok cloudprovider. Given that they aren't used through the critical path of code, is this something that we think that we need?

Reply:
FWIW it is something that we use in getOverrides for creating the launch template in karpenter-provider-aws. I think this looks a lot better than getting the zero indexed value from a slice on requirements every time.

Also agree that it doesn't have to be a panic and could be an error/fatal log.

Reply:
Still leaning toward fatal or panic myself, the main benefit IMO is enforcing the invariant rather than failing silently by returning an empty string. Since the offerings we generate are relatively static (they will always have the same requirements), this should never panic at runtime unless there's a programming error, in which case a panic will help ensure we catch it before release. Granted, since it's really only used in the provider and tests the benefit here is a little bit tenuous, but I'd rather have it than not IMO.